[Entity Analytics][Risk Engine] Risk Scoring Task

rylnd opened the pull request:
## What this PR does
* Adds a new Task Manager task, `risk_engine:risk_scoring`, responsible for invoking the `calculateAndPersistRiskScores` API defined in the risk scoring service.
  * Unlike an alerting task, we do not encrypt/persist an API key for the user. Instead, we use the internal kibana user to query all alerts in the current space.
  * The task configuration is stored as part of the existing `risk-engine-configuration` Saved Object
* Extends the `risk-engine-configuration` SO to include more configuration fields
  * Management of this configuration is not currently exposed to the user. They can only enable/disable the entire "Risk Engine" on the `Settings -> Entity Risk Score` page
  * The settings currently serve mainly as the "default" values for task execution, but also as a way for a customer/SA to modify task execution if necessary.
  * We expect to be modifying these default values before release, as part of our planned "tuning" stage.

### How to Review
* Setup:
  * The risk engine acts on Detection engine alerts, and so you will need to create:
      1. some "source" data (logs, filebeat, auditbeat, etc)
      2. Rules looking for the above "source" data, and generating alerts
  * The risk engine requires two feature flags, currently: `riskScoringPersistence` and `riskScoringRoutesEnabled`
  * You will also need a Platinum or greater license.
1. Test that the task executes correctly
    1. With the above data set up, navigate to `Settings -> Entity Risk Score` page, and enable the task by toggling `Entity risk scoring` to `On`
    1. Within a few minutes, risk scores should be written to the risk score datastream: 
        * `GET risk-score.risk-score-default/_search`
        * Replace `default` with the name of your current space, as necessary.
    1. Disabling/re-enabling the risk engine should trigger another execution of the task (similar to disabling/enabling a DE rule)
1. Enable the risk engine in another space
    * The engine (and task) can be enabled/executed in any kibana space.
    * Because the engine only acts upon alerts in the current space, you will need to first ensure alerts exist in that space.
1. Validate the data/mappings of persisted risk scores
    * Scores are based on the Stage 1 [ECS RFC](https://github.com/elastic/ecs/pull/2236)
    * There is no UI reading from these scores, currently (but that is introduced in https://github.com/elastic/kibana/pull/163237)
  
  

### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

rylnd committed:
WIP: defining the basics of our risk scoring task

* Adds the basic registration/start code paths, although we still need to
  `start()` the task from our management route
* Adds some of the task state, although I'm not entirely sure what we
  want/need there.
* Still need to pull most of the execution parameters from our new Saved
  Object.

rylnd committed:
RiskEngineDataClient uses scoped esClient

Rather than creating the transforms/indices on behalf of the internal
kibana user, we now do so on behalf of the user who initially set
up/enabled the risk engine.

If we have need for the internal user in the future, we can add that as
a separate dependency.

This also refactors some of the setup so that we instantiate the data
client when it is requested from the request context. We do not
currently have a need to create a data client outside of the context of
a user request.

rylnd committed:
Instantiate the RiskEngineDataClient with an SO client and the namespace

After a previous refactor we were no longer using the `user` argument to
a lot of these methods, and the `savedObjectsClient` and `namespace`
args were being passed around a bunch because they came from `init`.

Since we don't currently have a need to instantiate the client/resources
outside of the current space, both the namespace and the soClient are
now part of the client's constructor.

I left the namespace as an argument to a few calls from routes; this
should be identical to the property internal to the client, and I don't
quite know what it would mean/whether it would work if they were
different. The SO client, for example, is coded for the current
namespace.

rylnd committed:
Add some helpers for converting a relative date range to an absolute one

Most queries that use 'now' cannot be cached by elasticsearch.

Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/shard-request-cache.html

rylnd committed:
Add (and use) configuration from our saved object

* Adds helper methods for converting data from configuration
* Adds retrieval of SO config and inputs index data to both dataclient
  and riskScoreService.

rylnd committed:
Create our initial configuration SO with reasonable default values

This includes the alerts index for the current space.

rylnd committed:
WIP: starting task and writing an integration test

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

rylnd committed:
Add risk engine task to CI test runner

rylnd committed:
More accurate integration test name

rylnd committed:
Update integration test

We added attributes to the SO, here.

rylnd committed:
Trying to thread through our init route properly

I do not like the way this is structure, but right now I'm just trying
to execute the task so that I can TDD this.

However, local ES/Kibana errors are preventing me from running FTRs
locally, so I guess I'll wait for CI tonight.

rylnd committed:
Remove overriding ssl config in our integration tests

All our tests invoke this with ssl: true, but in the case where we're
trying to run tests against e.g. localhost this override prevents us
from being able to use a non-SSL connection.

rylnd committed:
WIP: Moving dependencies around to get enable/disable of task working

This all works, now, but the code is super confusing and I hate it. Will
be refactoring once integration tests are written.

rylnd committed:
Disable o11y plugin in our integration tests

What are the odds that we rely on behavior, there? Let's find out!

rylnd committed:
Better integration tests, and more OF them

* Renames a test file because it was identical to the production
  filename and driving me crazy
* Moves some shared logic into utils: SO management, invoking Risk
  Engine routes, etc.
* Excercises most of the functionality of the task via tests. I'll
  probably come up with some more here but I think this will be enough
  to confidently refactor the task code

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

rylnd committed:
Fix some basic CI checks

* Remove console statements
* lint improperly spaced comments :(

rylnd committed:
Fix flaky test by ensuring our `waitFor` is not a false positive

One (or more) of these tests would occasinally fail when the first batch
of 5 scores caused `waitForRiskScores` to return prematurely (before the
second batch of 5 scores was available in ES).

By specifying a scoreCount, we can address that race condition.

Closes #162736.

rylnd committed:
Test (and fix) removal of the risk scoring task

I couldn't find a permutation of `removeIfExists` that actually worked,
but now that this is under test I'll take another pass later.

rylnd committed:
Common description for all our risk engine integration tests

So that I can run `--grep "Risk Engine"` to run all these.

rylnd committed:
Remove security dependency from risk engine routes

We were previously using this to retrieve the username, but that
functionality no longer exists.

rylnd committed:
Clean up Risk Engine routes

* Remove unused dependencies
* Better error messages

rylnd committed:
Refactors task code into component functions

* Adds `interval` to SO configuration
* Adds `namespace` to task state so that a per-space task can be
  executed (theoretically)
* Updates tests

rylnd committed:
Add TODO

So's I don't forget

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

rylnd committed:
Update data client unit tests

* enable/disable now both take a taskmanager instance
* On enable we actually retrieve configuration twice: once to get
  configuration to invoke the task, and again to update the SO itself.
  Tests were updated accordingly.
* Remove unnecessary mockreset() call, as the mock itself is reassigned
  after every test.

rylnd committed:
Prevent an empty object/array filter arg from breaking our scoring query

This is the default value for our configuration filter, which we now
respect in our task.

rylnd committed:
Update risk engine configuration SO hash

We added some additional fields here corresponding to parameters for
risk score calculation (and one for the task interval).

rylnd committed:
Risk scoring task can now score both hosts and users

If the new SO configuration field, `identifierType`, is left undefined,
both 'hosts' and 'users' will be scored. Either value can be specified
in configuration to restrict scoring to just that type.

rylnd committed:
Address some race conditions in Risk Engine setup

* There is a race condition in the task initialization, between the task
  starting and the config SO being updated to `enabled: true`. If the
  task executes before the latter is made available, the task will short
  circuit that first time. It will run normally on subsequent
  executions, of course. This updates the order of initialization to
  make that less likely, but I believe it is still possible. I'm going
  to put these through the flaky runner to see.
* Also addresses an issue in test where the `disable` route's side
  effects (i.e. disabling the engine via the config SO) might not be
  immediately available. This fixes that flakiness in the test itself,
  although I think the "lag" is still going to be present in kibana.

rylnd committed:
Allow risk scoring task to be executed in multiple spaces

We register this one, but by scheduling via a space-specific task ID we
can run one in each space.

rylnd committed:
node scripts/check_mappings_update --fix

rylnd committed:
Update core snapshot hash for our risk configuration SO

We added a field to allow users to specify which entity types they want
to score in the task.

rylnd committed:
Add back our log option to this delete helper

The reason it wasn't using it was because we weren't using countDownTest
as the others do. Since that's more robust, I fixed it the correct way.

rylnd committed:
Allow soClient to retrieve configuration from the current space

Without this configuration, the soClient was unable to read
configuration created in a non-default space.

rylnd committed:
Add some more task logging

rylnd committed:
Merge branch 'main' into risk_engine_task

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/tasks/risk_scoring_task.ts:
```diff
@@ -0,0 +1,257 @@
+  const getRiskScoreService: GetRiskScoreService = (namespace) =>
+    getStartServices().then(([coreStart, _]) => {
+      const esClient = coreStart.elasticsearch.client.asInternalUser;
+      const soClient = coreStart.savedObjects.createInternalRepository(
+        undefined,
+        buildSpaceExtensions(namespace)
+      );
```
@elastic/kibana-core can someone please validate that this usage is correct? This task needs to be executed in an arbitrary kibana space, outside of the context of a request. I encode the `namespace` as part of the task's state during scheduling, and then the `soClient` eventually receives that knowledge here. Without [this code](dfdb10ab5749805d105d50190cfe3a90f14b4111), this internal SO client was unable to retrieve a saved object (namely the `risk-engine-configuration`) from a non-default space. It's also possible that the SO was created incorrectly to cause this behavior, but it looked correct `namespaces: ['example-space']` and this code now works.

rylnd replied:
@TinaHeiligers thank you for the review! Would you be able to take a look at this code as well? I couldn't find any examples of people using these extensions, and this was just the first combination of `getCurrentNamespace` and `getSearchableNamespaces` that worked for my purposes.

TinaHeiligers replied:
The @elastic/kibana-security team "owns" the extensions. @jeramysoucy can you 👁️ too please?
I'll review this PR.

TinaHeiligers replied:
@rylnd the code looks fine to me according to the `jsdocs` for extensions. When it comes to namespace handling, I think that's fine too. I've asked the Kibana security folks to take a look because I think your implementation is the first (or one of the first) outside of Core and Security domains.

rylnd replied:
Thank you @TinaHeiligers 🙏 . One more point that gave me pause here: the extensions docs make a point about "namespace ID" not being the same as a namespace string, and reference the corresponding `SavedObjectUtils`, but they are not explicit about which (strings or IDs) are passed as arguments to these extension functions, nor which are expected to be returned from them.

jeramysoucy replied:
@rylnd Let me make sure I understand the use case. This creates an SO client using a custom spaces extension that limits all operation to the current space. You're not using the security extension or encryption extension. You're using this SO client to create/update the configuration for the risk engine. You rely on the find function to retrieve the config object for the current space. Is this all accurate? Anything missing?

Is there a reason you couldn't use the existing spaces extension? It is designed to operate only within the active space, and you can limit which searched spaces with the find function options.

I will be on PTO next week so I am tagging @azasypkin and @legrego in case you need to follow up on this sooner.

rylnd replied:
@jeramysoucy thanks for the 👀 ! 

> This creates an SO client using a custom spaces extension that limits all operation to the current space

Nope, this client is only used to _retrieve_ the SO; the SO itself is created with the scoped SO client from the request context. The use case is simply "retrieve our configuration SO for the current space," and when I was using the client returned from an empty `createInternalRepository()` call, it was not able to retrieve that SO in a non-default space.

As I mentioned above, the other possibility is that the SO itself is somehow [created](https://github.com/elastic/kibana/pull/163216/files#diff-dde99144e175122a90c2e43b30bd9ea91034597c44e85fa296b78ba9786cf022R90-R94) improperly, but since I was able to find it with this extension that seems less likely? 

If you're saying that this _should_ work without these extensions (is that what you're saying?), I'll dig deeper next week. I would definitely prefer not to use these extensions, and certainly not a subset of them, if possible.

rylnd replied:
@jeramysoucy @azasypkin @legrego I have not had luck using `createInternalRepository` during kibana setup to obtain the SO we need for the particular space at task runtime. Logically this makes sense, since during kibana setup there isn't really a concept of "current space" (or it's simply the default space). 

I can use anything from the `savedObjects` plugin to accomplish this task, but I do not have a `request` object from which to derive a scoped client (since that would require serialization of the request); I only have a namespace string. Is there a better way to accomplish my goal, or are extensions the way to do this? Any assistance would be appreciated.

rylnd replied:
Thanks for all the help here, everyone! After discussion with both @azasypkin and @legrego I opted to go with the (fake) request-based route for instantiating a scoped SO client: [64d9cf5](https://github.com/elastic/kibana/pull/163216/commits/64d9cf59948b0e26b6676ffd1e306b439525e630). 

Things look to be good (as evidenced by smoke testing and a [new integration test](https://github.com/elastic/kibana/pull/163216/commits/c8c1c12a254d152264de70f43a069158a73d51a8)), but please give this a once over and confirm that things look good on your end!

azasypkin replied:
Considering that we use the SO client without the security extension only within a background task and solely for a limited and seemingly safe set of operations, looks good to me. I just left one comment/ask.

rylnd commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into risk_engine_task

rylnd committed:
Merge branch 'main' into risk_engine_task

 Conflicts:
	x-pack/plugins/security_solution/common/risk_engine/index.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.test.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/routes/risk_engine_disable_route.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/routes/risk_engine_enable_route.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/routes/risk_engine_init_route.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/types.ts
	x-pack/plugins/security_solution/server/lib/risk_engine/utils/saved_object_configuration.ts
	x-pack/plugins/security_solution/server/plugin.ts
	x-pack/test/detection_engine_api_integration/security_and_spaces/group10/risk_engine/init_and_status_apis.ts

rylnd committed:
Merge branch 'main' into risk_engine_task

rylnd committed:
Fix bad resolution

Git auto-merged this one for us. Thanks, git!

rylnd committed:
Merge branch 'main' into risk_engine_task

TinaHeiligers reviewed (approved):
Core mappings and saved objects migration map changes look ok to me.

rylnd committed:
Merge branch 'main' into risk_engine_task

rylnd committed:
Merge branch 'main' into risk_engine_task

nkhristinin reviewed (changes requested):
Thanks @rylnd, for this work!

Finally, review it and test locally: The task is working, and scores are generated.
I tested with different spaces, and stop/run kibana - all works well.

I think we definitely should address the visibility of tasks in future work because without that it's quite hard to understand what is happening with the risk engine.

I left a couple of comments, most of them just small suggestions, but there is one which I think we should change - it is about throwing an error from the run/stop task, to be able to catch it on a level above, and reflect it in API calls

rylnd commented on packages/kbn-check-mappings-update-cli/current_mappings.json:
```diff
@@ -2953,8 +2953,34 @@
+      "pageSize": {
+        "type": "integer"
+      },
+      "range": {
```
`range` is (and has always been) a convenience field for the most common filtering case (by date range). Consolidating to a single `filter` property would require API calls (and those editing the saved object) to manually type out the full filter. 

I view `filter` as the "low level, high flexibility" option, but we have `range` and `identifier_type` as more common, specific ways to refine the scoring.

rylnd replied:
@nkhristinin what do you think?

rylnd commented on x-pack/test/detection_engine_api_integration/security_and_spaces/group10/risk_engine/utils.ts:
```diff
@@ -112,29 +119,115 @@ export const deleteAllRiskScores = async (
-  log: ToolingLog,
-  index: string[] = ['risk-score.risk-score-default']
-): Promise<void> => {
+export const waitForRiskScoresToBePresent = async ({
```
It _can_ be used to wait for risk scores from the `latest` transform, if you specify the appropriate `index` parameter. I don't think `waitForRiskScoresToBeWritten`, `waitForRiskScoresToBeReadable`, nor `waitForRiskScoresFromTask` are any more clear in regards to the `latest` complication...

I'm happy to make a change if you can think of a better name; for now I'll add a jsdoc comment to clarify 👍 .

rylnd replied:
Addressed in [672fd00](https://github.com/elastic/kibana/pull/163216/commits/672fd00e2127eac5f79963921edd2be975373f6f)

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/tasks/risk_scoring_task.ts:
```diff
@@ -0,0 +1,257 @@
+
+  updatedState.scoresWritten = scoresWritten;
+
+  log('task run completed');
```
Yep, the `log` function here is a decorator around `logger.info` that adds the task ID: https://github.com/elastic/kibana/pull/163216/files#diff-9044d1687550dc849d2e459598580ddb16c438557d31d34e8b772c90bc58f7dbR39

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/tasks/risk_scoring_task.ts:
```diff
@@ -0,0 +1,257 @@
+  } catch (err) {
+    if (!SavedObjectsErrorHelpers.isNotFoundError(err)) {
+      logger.error(`Failed to remove risk scoring task: ${err.message}`);
+      throw err;
```
This is being re-thrown since it's called (almost) directly from the `_disable` API, which will catch this error and return it as a 400.

rylnd replied:
I assume you're not requesting a change here, just pointing out the inconsistency?

nkhristinin replied:
re-trhow appear only in "if" block, can be there a case that we have other type of error, not going in the "if", and then silently doing nothing?

rylnd committed:
Add JSDoc for some ambiguously named test helper functions

rylnd committed:
Merge branch 'main' into risk_engine_task

rylnd committed:
Re-throw errors from taskmanager when enabling the risk scoring task

This way the UI/API can convey it to the user.

rylnd committed:
Add some tests around enable/disable routes

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/tasks/risk_scoring_task.ts:
```diff
@@ -0,0 +1,257 @@
+      params: { version: VERSION },
+    });
+  } catch (e) {
+    logger.warn(`[task ${taskId}]: error scheduling task, received ${e.message}`);
```
addressed in [2879d8a](https://github.com/elastic/kibana/pull/163216/commits/2879d8afc7870766843b9dd978b384287b735311)

rylnd committed:
Better synchronize enabling of risk engine task

If an error is raised during enablement (either updating the SO or
starting the task), we will undo enablement by disabling the engine.

rylnd commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_engine_data_client.ts:
```diff
@@ -133,25 +140,49 @@ export class RiskEngineDataClient {
       attributes: {
         enabled: true,
       },
     });
```
@nkhristinin I didn't implement a better way to synchronize things, but I did implement a "reversal" if enablement fails in [7cf14b9](https://github.com/elastic/kibana/pull/163216/commits/7cf14b97fb423763c9903f519663ac0f7f0d8b02).

rylnd committed:
Merge branch 'main' into risk_engine_task

nkhristinin reviewed (approved):
Thanks for addressing the comments, LGTM!

rylnd committed:
linting: remove redundant await

rylnd committed:
Add risk engine test for non-default space

This required threading through a `namespace` parameter for a few rule
helpers, but I made it optional so it should be a minimal change.

rylnd committed:
Remove outstanding TODO

We do not need to return a schedule here until we have need to
override the default interval. When the risk engine is enabled,
the task is scheduled with the interval configured in the saved object.

rylnd committed:
Use a more robust SavedObjects client in our risk scoring task

Rather than defining the SO extensions ourselves, we instead build a
fake kibana request for the particular namespace we're in, and use the
request-based helpers to build a scoped SO client from that request. In
this way, we're not coupled to the extensions implementation, and
additionally we have the encryption and spaces extensions
defined/enabled for us.

We need to disable the security extension, however, because there is not
a kibana system user that would be able to access these Saved Objects,
normally.

rylnd committed:
Merge branch 'main' into risk_engine_task

 Conflicts:
	src/core/server/integration_tests/saved_objects/migrations/group2/check_registered_types.test.ts

rylnd committed:
Add persistence feature flag to our cypress tests

Now that we're actually enabling the task when we enable the risk
engine, we need this flag enabled, or else we'll get an error because
the task wasn't registered on kibana startup.

rylnd committed:
Remove unnecessary `visit` step from our test

We can navigate directly to the entity analytics management page; we
don't need to visit alerts for these tests.

azasypkin commented on x-pack/plugins/security_solution/server/lib/risk_engine/tasks/helpers.ts:
```diff
@@ -0,0 +1,66 @@
+  return request;
+};
+
+export const buildScopedInternalSavedObjectsClient = ({
```
__nit:__ Could you please consider adding a comment for future readers to explain the purpose of this method, the reason for disabling the security extension, and why this method should be used sparingly (e.g., only in scenarios where there's no access to the user context, such as within a background task)? While it may seem obvious now, it's easy to forget these details after a few months, and the comment can serve as a helpful reminder. Additionally, it's our practice to prefix such methods with `Unsafe` or `Insecure` to clearly indicate that they should be used with caution.
```suggestion
export const buildScopedInternalSavedObjectsClientUnsafe = ({
```

rylnd replied:
Will do! I still have the habit of putting that context in commit messages, but our squashing strategy makes that much less useful 😅 .

rylnd replied:
Addressed in https://github.com/elastic/kibana/pull/163216/commits/80785e2ee3ec48b99e5d70be5956625dacd11d95.

rylnd committed:
Rename and document function to make its purpose more clear

rylnd committed:
Revert integration test config changes

I didn't quite understand what these SSL changes were doing, and
although they seem to be unused I'm just not going to touch them for
now. This was committed in the course of running some things locally
against a non-SSL kibana, but that time has passed.

Additionally, this reverts the disabling of observability in integration
tests. While this was useful for developing locally, it's not something
that should be in main.

rylnd committed:
Merge branch 'main' into risk_engine_task

 Conflicts:
	x-pack/test/detection_engine_api_integration/utils/create_rule.ts
	x-pack/test/detection_engine_api_integration/utils/wait_for_rule_status.ts

rylnd committed:
Update pending tests

* Rescheduling is inherent to the fact that define an `interval` for our
  schedule; no need to test that explicitly
* Also removes tests around failure conditions; the entire task run is
  wrapped in a try/catch such that we don't need to do that ourselves.

rylnd committed:
Merge branch 'main' into risk_engine_task

rylnd committed:
Merge branch 'main' into risk_engine_task
